source/MaterialXGenShader/ShaderTranslator.cpp

```diff
@@ -0,0 +1,207 @@
+{
+    DocumentPtr doc = shaderRef->getDocument();
+    vector<OutputPtr> outputs = doc->getNodeGraph("NG_" + _translationNode->getCategory())->getOutputs();
+    for (OutputPtr translationGraphOutput : outputs)
+    {
+        // Updating the shaderref sockets
+        string outputName = translationGraphOutput->getName();
+        outputName = outputName.substr(0, outputName.find("_out"));
+        OutputPtr translatedOutput = _graph->addOutput(outputName + "_out", translationGraphOutput->getType());
+        BindInputPtr translatedBindInput = shaderRef->addBindInput(outputName, translationGraphOutput->getType());
+        translatedBindInput->setConnectedOutput(translatedOutput);
+        // if normals need to be transformed into world space
+        if (connectsToNormalMapNode(translationGraphOutput))
+        {
+            insertOutputNgUpstreamElems(translatedOutput, translationGraphOutput);
+        }
+        else
+        {
+            // registering outputs from translation node
+            NodePtr outNode = _graph->addNode("dot", outputName + "_dot", translationGraphOutput->getType());
+            translatedOutput->setConnectedNode(outNode);
+
+            InputPtr dotNodeInput = outNode->addInput("in", translationGraphOutput->getType());
+
+            // if value does not need to be computed
+            if (translationGraphOutput->getNodeName() == EMPTY_STRING)
+            {
+                dotNodeInput->setValueString(translationGraphOutput->getValueString());
+            }
+            else
+            {
+                dotNodeInput->setConnectedNode(_translationNode);
+                dotNodeInput->setOutputString(outputName + "_out");
+            }
+        }
+    }
+}
+
+void ShaderTranslator::translateShader(ShaderRefPtr shaderRef, string destShader)
+{
+    if (!shaderRef)
+    {
+        return;
+    }
+
+    DocumentPtr doc = shaderRef->getDocument();
+    string translateNodeString = shaderRef->getNodeString() + "_to_" + destShader;
+    if (!_translationNodes.count(translateNodeString))
+    {
+        return;
+    }
+    _graph = doc->getNodeGraph(shaderRef->getBindInputs()[0]->getAttribute(PortElement::NODE_GRAPH_ATTRIBUTE));
+    _translationNode = _graph->addNode(translateNodeString, "translation", MULTI_OUTPUT_TYPE_STRING);
+
+    connectToTranslationInputs(shaderRef);
+    shaderRef->setNodeString(destShader);
+    connectTranslationOutputs(shaderRef);
+}
+
+bool ShaderTranslator::translateAllMaterials(DocumentPtr doc, string destShader)
+{
+    ShaderTranslatorPtr translator = ShaderTranslator::create(doc);
+    vector<TypedElementPtr> renderableShaderRefs;
+    std::unordered_set<OutputPtr> outputs;
+    findRenderableShaderRefs(doc, renderableShaderRefs, false, outputs);
+    for (TypedElementPtr elem : renderableShaderRefs)
+    {
+        ShaderRefPtr sr = elem ? elem->asA<ShaderRef>() : nullptr;
+        string startShader = sr->getNodeString();
```

Comment:
It might be clearer to name these variables "sourceShader" and "sourceName", making the common parallel with "destShader" above.